drivers/include/periph/gpio_ll.h

```diff
@@ -0,0 +1,498 @@
+ *          will set the bits of non-output pins as needed to not have side
+ *          effects on the state the GPIO port had when calling
+ *          @ref gpio_ll_prepare_write .
+ *
+ * @param[in] port      port to modify
+ * @param[in] value     new state of the GPIO port to write
+ *
+ * Usage:
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.c}
+ *  // emit a square wave from two pins with a phase shift of pi between the
+ *  // waves, e.g.:
+ *  // pin1: _⎍_⎍_⎍_⎍_⎍_⎍_⎍_⎍_⎍
+ *  // pin2: ⎍_⎍_⎍_⎍_⎍_⎍_⎍_⎍_⎍_
+ *  void square_wave(gpio_port_t port, uint8_t pin1, uint8_t pin2)
+ *  {
+ *      uword_t state1 = gpio_ll_prepare_write(port, 1 << pin1);
+ *      uword_t state2 = gpio_ll_prepare_write(port, 1 << pin2);
+ *      while (1) {
+ *          gpio_ll_write(port, state1);
+ *          gpio_ll_write(port, state2);
+ *      }
+ *  }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * @note    If the configuration of the used port changes between the calls
+ *          of @ref gpio_ll_prepare_write and @ref gpio_ll_write the write will
+ *          still have undesired side effects on the configuration of input pins
+ *          if the platform multiplexes the write register with configuration
+ *          for input pins. It is expected that the user of the API either
+ *          exclusively uses a GPIO port or synchronizes with other users to
+ *          update the prepared value on configuration changes.
+ */
+static inline void gpio_ll_write(gpio_port_t port, uword_t value);
+
+/**
+ * @brief   Extract the `gpio_port_t` from a `gpio_t`
+ */
+static inline gpio_port_t gpio_get_port(gpio_t pin);
+
+/**
+ * @brief   Extract the pin number from a `gpio_t`
+ */
+static inline uint8_t gpio_get_pin_num(gpio_t pin);
+
+/**
+ * @brief   Pack a pointer into a @ref gpio_port_t
+ *
+ * @pre     The address in @p addr is not `NULL` ***and*** points to a regular
+ *          memory region (e.g. anywhere in `.data`, `.rodata`, `.bss` or an
+ *          address returned by `malloc()` and friends)
+ *
+ * @return  A value that is distinguishable from any valid port reference and
+ *          from which @p addr can be reconstructed by calling
+ *          @ref gpio_port_unpack_addr
+ */
+static inline gpio_port_t gpio_port_pack_addr(void *addr);
+
+/**
+ * @brief   Extract a data pointer that was packed by @ref gpio_port_pack_addr
+ *
+ * @return  The exact address previously packed by @ref gpio_port_pack_addr
+ * @retval  NULL    @p port is a valid GPIO port
+ *
+ * The motivation is that a high level API can multiplex peripheral GPIOs and
+ * GPIOs provided by port extenders. That high level API could pack pointers to
+ * the device descriptors of port extenders into a @ref gpio_port_t and use this
+ * function to check if the port is a peripheral port (the return value is
+ * `NULL`), or retrieve the device descriptor.
```

Comment:
Ah now this is becoming more clear.
It's more like a `bool gpio_addr_is_port(const void *addr)` type of function

Reply:
I couldn't verify if this really works. To get proof of this, it would be necessary to implement `gpio_ll` for a GPIO expander. Since we don't have that at the moment, we should leave it that way for now.